### PR TITLE
Max gas limit

### PR DESCRIPTION
**Pull request by wuminzhe**

### CODE REVIEW

File: contracts/periphery/contracts/s2s/calls/PangolinCalls.sol [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+library PangolinCalls {
+    using SafeMath for uint256;
+
+    // According to the EVM gas benchmark, 1 gas ~= 40_000 weight.
+    uint64 public constant WEIGHT_PER_GAS = 40_000;
+
+    // The cap limitation for gas limit
+    // https://github.com/darwinia-network/darwinia-messages-substrate/issues/107#issuecomment-1164185135
+    uint64 public constant MAX_GAS_LIMIT  = 100_000_000_000;
```
Comment (hujw77) [inline contracts/periphery/contracts/s2s/calls/PangolinCalls.sol:17]: Gas limit seems to too large？

**Review by hujw77 — COMMENTED** [review]
how to benchmark the MAX_GAS_LIMIT

File: contracts/periphery/contracts/s2s/calls/PangolinCalls.sol (hunk 1 above)
Comment (wuminzhe) [reply contracts/periphery/contracts/s2s/calls/PangolinCalls.sol:17]: I get it from rust code. not right?

https://github.com/darwinia-network/darwinia-common/blob/eaf8704a2b40d4facb800cc98852772b144de346/node/runtime/pangolin/src/pallets/bridge_dispatch.rs#L26

Comment (hujw77) [reply contracts/periphery/contracts/s2s/calls/PangolinCalls.sol:17]: GAS !== Balance

Comment (wuminzhe) [reply contracts/periphery/contracts/s2s/calls/PangolinCalls.sol:17]:
```
gas_limit * min_gas_price should <= maxUsableBalance * 10**9
gas_limit                 should <= maxUsableBalance * 10**9 / min_gas_price
gas_limit                 should <= maxUsableBalance * 10**9 / 1 GWEI
gas_limit                 should <= 100 * COIN * 10**9 / 1 GWEI
gas_limit                 should <= 100 * 10**9 * 10**9 / 1 GWEI

so,
max_gas_limit = 100 * 10**9
```

Comment (hujw77) [reply contracts/periphery/contracts/s2s/calls/PangolinCalls.sol:17]: I am confused about this, Please help to review @HackFisher @xiaoch05

Comment (xiaoch05) [reply contracts/periphery/contracts/s2s/calls/PangolinCalls.sol:17]: fee = gaslimit * gasprice = 100 COIN
here the gaslimit is too large, usually about several million, can't exceed 2000w in ethereum network.

Comment (hujw77) [reply contracts/periphery/contracts/s2s/calls/PangolinCalls.sol:17]: COIN = 10**9
Is there a precision here?

**Comment by HackFisher** [comment]
I think we can not limit only by gas_limit here, the gas fee equals gas_limit * gas_price should not larger than 100 CRAB (current settings), 1 CRAB equals 1**18 ether on Smart chain, and 1 ** 9 nano on Crab substrate chain, notorious different decimal we all know.

For here, since we do not know the gas_price when block will be included, we can not infer the exact gas_limit then.


This limit can be documented for gas senders to note, can not be hard coded in SDK IMO.

I'm not sure it is a sample here or part of SDK.

**Comment by hujw77** [comment]
> @hujw77 What's the maximum block gas limit then?

https://github.com/darwinia-network/darwinia/blob/03c86e1132dce1f6de08decfde2f8a536dd65e79/runtime/crab/src/pallets/evm_.rs#L167

File: contracts/periphery/contracts/s2s/calls/PangolinCalls.sol [hunk 2]
```diff
@@ -12,6 +12,10 @@ library PangolinCalls {
     // According to the EVM gas benchmark, 1 gas ~= 40_000 weight.
     uint64 public constant WEIGHT_PER_GAS = 40_000;
 
+    // The cap limitation for gas limit
+    // https://github.com/darwinia-network/darwinia-messages-substrate/issues/107#issuecomment-1164185135
+    uint64 public constant MAX_GAS_LIMIT  = 100_000_000_000;
```
Comment (HackFisher) [inline contracts/periphery/contracts/s2s/calls/PangolinCalls.sol:17]:
```suggestion
    // The cap limitation for gas limit, we use 1/5 of max block gas limit 
    // The minimum limitations of max block gas limit and considerations of MaxUsableBalanceFromRelayer  fee limit
    // https://github.com/darwinia-network/darwinia-messages-substrate/issues/107#issuecomment-1164185135
    uint64 public constant MAX_GAS_LIMIT  = 10_000_000;
```

**Review by HackFisher — APPROVED** [review]

**Pull request merged**